R/step_epi_slide.R

```diff
@@ -0,0 +1,277 @@
+#'  2. A character string name of a function, e.g., `"mean"`. Note that this
+#'     can be difficult to examine for mistakes (so the misspelling `"maen"`
+#'     won't produce an error until you try to actually fit the model)
+#'  3. A base `R` lambda function, e.g., `function(x) mean(x, na.rm = TRUE)`
+#'  4. A new-style base `R` lambda function, e.g., `\(x) mean(x, na.rm = TRUE)`
+#'  5. A one-sided formula like `~ mean(.x, na.rm = TRUE)`.
+#'
+#'  Note that in cases 3 and 4, `x` can be any variable name you like (for
+#'  example `\(dog) mean(dog, na.rm = TRUE)` will work). But in case 5, the
+#'  argument must be named `.x`. A common, though very difficult to debug
+#'  error is using something like `function(x) mean`. This will not work.
+#' @param f_name a character string of at most 20 characters that describes
+#'   the function. This will be combined with `prefix` and the columns in `...`
+#'   to name the result using `{prefix}{f_name}_{column}`. It will be determined
+#'   automatically using `clean_f_name()`.
+#' @param before,after non-negative integers.
+#'   How far `before` and `after` each `time_value` should
+#'   the sliding window extend? Any value provided for either
+#'   argument must be a single, non-`NA`, non-negative,
+#'   [integer-compatible][vctrs::vec_cast] number of time steps. Endpoints of
+#'   the window are inclusive. Common settings:
+#'   * For trailing/right-aligned windows from `time_value - time_step(k)` to
+#'   `time_value`, use `before=k, after=0`. This is the most likely use case
+#'   for the purposes of forecasting.
+#'   * For center-aligned windows from `time_value - time_step(k)` to
+#'   `time_value + time_step(k)`, use `before=k, after=k`.
+#'   * For leading/left-aligned windows from `time_value` to
+#'   `time_value + time_step(k)`, use `after=k, after=0`.
+#'
+#'   You may also pass a [lubridate::period], like `lubridate::weeks(1)` or a
+#'   character string that is coercible to a [lubridate::period], like
+#'   `"2 weeks"`.
+#' @template step-return
+#'
+#' @export
+#' @examples
+#' library(dplyr)
+#' jhu <- case_death_rate_subset %>%
+#'   filter(time_value >= as.Date("2021-01-01"), geo_value %in% c("ca", "ny"))
+#' rec <- epi_recipe(jhu) %>%
+#'   step_epi_slide(case_rate, death_rate,
+#'     .f = ~ mean(.x, na.rm = TRUE),
+#'     before = 6L
+#'   )
+#' bake(prep(rec, jhu), new_data = NULL)
+step_epi_slide <-
+  function(recipe,
+           ...,
+           .f,
+           before = 0L,
+           after = 0L,
+           role = "predictor",
+           prefix = "epi_slide_",
+           f_name = clean_f_name(.f),
+           skip = FALSE,
+           id = rand_id("epi_slide")) {
+    if (!is_epi_recipe(recipe)) {
+      rlang::abort("This recipe step can only operate on an `epi_recipe`.")
+    }
+    .f <- validate_slide_fun(.f)
+    arg_is_scalar(before, after)
+    before <- try_period(before)
+    after <- try_period(after)
```

Comment:
This is probably a case where `:::` makes sense, since I doubt we really want to make `validate_slide_window_arg` public, and we actually are the devs on both packages.

Reply:
It can be exported but not "public". Add the `#' @noRd` tag. You can also add `#' @keywords internal`. Then no documentation will be generated and it won't appear on the pkgdown site.

Reply:
Ah yea ok, if this PR changes to use epiprocess::epi_slide or epiprocess::epi_slide_opt, then I'd definitely recommend reusing `validate_slide_window_arg` with `attributes(epi_df)$metadata$time_type` here and removing `try_period`.